### ISSUE TITLE
Intermittent Failures to Deliver OTR Messages

### ISSUE THREAD

**Issue by mddvul22** [RocketChat/Rocket.Chat#28918]
### Description:

We have just upgraded to Rocket Chat 6.0.2.  With this upgrade, we are now seeing random (but consistent) failures in delivering OTR messages.  After an OTR session is established, some messages are delivered while others are not.

### Steps to reproduce:

1. Establish an OTR session
2. Send repeated messages
3. Some messages will stay gray in color and the other user will never see those messages.

### Expected behavior:

I would expect delivery of every message sent during the OTR session.

### Actual behavior:

Some messages are delivered, and some are not.  The screenshot, below, shows two different logged in users side-by-side in an OTR session.  The user on the left is sending all of the OTR messages.  Only some are getting delivered, to the user on the right.

![image](https://user-images.githubusercontent.com/6833993/231887056-26e95f29-7a5d-488f-ab7d-5cc5047e18c4.png)

### Server Setup Information:

- Version of Rocket.Chat Server: 6.0.2
- Operating System: Red Hat Linux 8.7
- Deployment Method: build images provided at https://cdn-download.rocket.chat/build/rocket.chat-6.0.2.tgz
- Number of Running Instances: 2
- DB Replicaset Oplog: enabled
- NodeJS Version: 14.21.2
- MongoDB Version: 4.4.19

### Client Setup Information

- Desktop App or Browser Version: Firefox/Chrome
- Operating System: Linux

### Relevant logs:

Not sure what logs we can provide for an OTR session.  However, if you think that we can get helpful logs for an OTR session, we can raise our logging level to try and get something to add here.

**Comment by mddvul22** [comment]
Yes, I have had reports from some of our employees of problems with OTR session setup failing, also.  There are definitely issues with online presence being incorrect, also.  And to be clear, in case anybody want to know, here is our license type:

![image](https://user-images.githubusercontent.com/6833993/232050175-a9ccf43c-e18d-4efc-a74e-24a1eb097cbf.png)

**Comment by wfleurant** [comment]
OTR is broken -- is it something to do with status updates also broken? re: both parties must be online for OTR to work?

**Comment by Gummikavalier** [comment]
> OTR is broken -- is it something to do with status updates also broken? re: both parties must be online for OTR to work?

I too think that a lot of RC6.x issues may be related to the timely handling of user statuses.

**Comment by jdsatava127** [comment]
We are experiencing this as well after upgrading to 6.x.

**Comment by mddvul22** [comment]
Yes, we were told that 6.2.3 would fix this issue, but it has not.  We upgraded to 6.2.3 last night, and the problem persists.

**Comment by someadmin** [comment]
I got the similar issue with thread messages, but without high CPU/memory consumprtion. It's after upgrading to 6.2.3 then to 6.2.5.
And records contained "Slow query" in mongodb logs started to appear.

**Comment by fcoppolani** [comment]
> > OTR is broken -- is it something to do with status updates also broken? re: both parties must be online for OTR to work?
> 
> I too think that a lot of RC6.x issues may be related to the timely handling of user statuses.

Hi,  Yes I just tested this, we were at 230 active sessions : no user status, no OTR, after a while I reactivate the user status at 199 sessions, OTR works ....
So great, not enough we had to cope with this breaking change (and the sentence "_ENTERPRISE: Limit presence statuses to 200 concurrent users when running monolith to keep performance_" is a perfect example of misleading communication) moreover there is a bug with OTR because of that limitation.... I just had an interview with rocket sales, and really I'm not sure I'm ready to pay the price you ask for so little reliability.
Regards,

**Comment by benedikt-wegmann** [comment]
Oh, and if you want to patch out that limitation, this is the file and line to do so: https://github.com/RocketChat/Rocket.Chat/blob/916c0dcaf22b2d891d2a257c8dc558f7768d6116/ee/packages/presence/src/Presence.ts#L9

**Comment by ankar84** [comment]
We have now same problem too!
We upgraded to 6.2.10 and faced that presence limit problem.
We have about 6000 active users, so now we must have about 40 instances to be able use that basic presence feature.
Actually, someone from Rocket.Chat team advised us to have about 12-15 instances to get best performance because in you have more instances you don't have performance benefits anymore. 
And now we are in trap - we must make more instances to patch Docker image.
cc @sampaiodiego @engelgabriel @marceloschmidt @rodrigok

**Comment by benedikt-wegmann** [comment]
Hey @ankar84 ! We went through the same experience.

According to our research (in the source code), the "200 active users" limit is *overall*, not per instance. Scaling up won't solve the issue.

Regarding your licence: you probably have a so-called "pro" licence in reality, not an enterprise license, although Rocket.Chat does display it as such (it did for us). This was a licence offered in the past, between the free community licence and the enterprise licence. You probably have features like the push notifications and and omnichannel but not this "engagement dashboard" and "device management"
Contact Rocket.Chat support, tell them GWDG (from Germany) sent you and ask them to convert your license to an actual EE license, because the recent update has crippled your service artificially. After the upgrade of your license you have to login as admin, go to Administration > Registration > Manage > Sync and then restart every instance that does not have `DISABLE_PRESENCE_MONITOR=false` set (i.e. every instance that handles user presence). You will see the user presence return immediately after the restarts.

If you want to go the "other" route you can patch this simple check in the source code, see my comment above for the relevant location in the code. :wink:

**Comment by Gummikavalier** [comment]
OTR seems to be working better for RC 6.5.0. But am I correct that enabling it and sending one message, it still creates ~700 MHz CPU usage on the server and client ends, and constant session with 400 kb/s between the participants?

Is this by design to create noise into the session?

**Comment by Gummikavalier** [comment]
To me it looks like that this one is now fixed in RC 6.5.1. We don't spot slowness, missing messages or high CPU usage by client or server ends anymore using OTR.

Likely related:

https://github.com/RocketChat/Rocket.Chat/commit/c2b224fd82f655f4edf6b4a7d5baea2c1166978a: Exceeding API calls when sending OTR messages

https://github.com/RocketChat/Rocket.Chat/pull/31019: fix: OTR calling a deprecated method

**Comment by mddvul22** [comment]
I also have a bug open with enterprise paid support that has been open for about 8 months on this issue.  If it is _really_ fixed, hopefully RC paid support will let me know.

**Comment by mddvul22** [comment]
We can confirm that the intermittent failure to deliver OTR messages inside an established OTR session is fixed, after we have upgraded to 6.5.2!
